discovery.go

```diff
@@ -0,0 +1,369 @@
+	ongoing map[string]struct{}
+
+	// done handles completion of a discovery request
+	done chan string
+
+	// peerNotifier processes all peer notifications
+	peerNotifier chan peer.ID
+
+	// tracks the channels subscribed to new peer notifications
+	newPeerNotif map[chan peer.ID]*peerEventTracker
+
+	// setup new peer notifier
+	addPeerNotif chan *reqNewPeerNotifier
+
+	// remove peer notifier
+	rmPeerNotif chan chan peer.ID
+}
+
+// newDiscover creates a new peer discovery mechanism
+func newDiscover() *discover {
+	return &discover{
+		advertising:  make(map[string]context.CancelFunc),
+		bootstrapped: make(map[string]chan struct{}),
+		discoverQ:    make(chan *discoverReq, 32),
+		ongoing:      make(map[string]struct{}),
+		done:         make(chan string),
+		peerNotifier: make(chan peer.ID, 32),
+		newPeerNotif: make(map[chan peer.ID]*peerEventTracker),
+		addPeerNotif: make(chan *reqNewPeerNotifier),
+		rmPeerNotif:  make(chan chan peer.ID),
+	}
+}
+
+// Start attaches the discovery pipeline to a pubsub instance and starts event loop
+func (d *discover) Start(p *PubSub) {
+	if d.discovery == nil {
+		return
+	}
+
+	d.p = p
+	go d.discoverLoop()
+}
+
+func (d *discover) discoverLoop() {
+	for {
+		select {
+		case discover := <-d.discoverQ:
+			topic := discover.topic
+
+			d.bootstrappedMux.RLock()
+			bootstrapDone := d.bootstrapped[topic]
+			d.bootstrappedMux.RUnlock()
+			select {
+			case <-bootstrapDone:
+			default:
+				continue
+			}
+
+			if _, ok := d.ongoing[topic]; !ok {
+				d.ongoing[topic] = struct{}{}
+				go func() {
+					d.handleDiscovery(d.p.ctx, topic, discover.opts)
+					d.done <- topic
+				}()
+			}
+		case topic := <-d.done:
+			delete(d.ongoing, topic)
+		case add := <-d.addPeerNotif:
+			ch := make(chan peer.ID, 10)
+			sentPeers := &peerEventTracker{peers: make(map[peer.ID]struct{})}
+			d.newPeerNotif[ch] = sentPeers
+			add.resp <- ch
+			go func() {
+				topicPeers := d.p.ListPeers(add.topic)
+				sentPeers.mux.RLock()
+				sendPeers := peerIntersect(topicPeers, sentPeers.peers)
+				sentPeers.mux.RUnlock()
+
+				sentPeers.mux.Lock()
+				for _, pid := range sendPeers {
+					if _, ok := sentPeers.peers[pid]; !ok {
+						sentPeers.peers[pid] = struct{}{}
+						ch <- pid
+					}
+				}
+				sentPeers.mux.Unlock()
+			}()
+		case rm := <-d.rmPeerNotif:
+			delete(d.newPeerNotif, rm)
+		case pid := <-d.peerNotifier:
+			d.notifyNewPeer(pid)
+		case <-d.p.ctx.Done():
+			return
+		}
+	}
+}
+
+// Advertise advertises this node's interest in a topic to a discovery service
+func (d *discover) Advertise(topic string) {
+	if d.discovery == nil {
+		return
+	}
+
+	advertisingCtx, cancel := context.WithCancel(d.p.ctx)
+
+	if _, ok := d.advertising[topic]; ok {
+		return
+	}
+	d.advertising[topic] = cancel
+
+	go func() {
+		next, err := d.discovery.Advertise(advertisingCtx, topic)
+		if err != nil {
+			log.Warningf("bootstrap: error providing rendezvous for %s: %s", topic, err.Error())
```

Comment:
shouldn't we short-circuit here as well?

Reply:
No? This seems right.